CPU Miner Enhancements

cculianu opened the pull request:
#### Description

I have been mining testnet lately (hashrate seems to be mostly CPU now anyway on testnet, with the occasional ASIC miner showing up to rain blocks for 12 hours before leaving abruptly).

As such, I have been using the `bitcoin-miner` program quite a bit and I wanted to add a few features to it to make it more useful for everybody.  I already am running this modified version on my setup and it's working great, but I figure I'd give back and share my "advances" with others.

This PR adds the following:

- `-maxdifficulty=<float>` command-line argument to `bitcoin-miner`. With this option, the miner doesn't bother mining if difficulty is too high. It's pointless to mine with a CPU when difficulty is too high (extra wear & tear on the hardware).  Whenever difficulty is > `maxdifficulty` it just sleeps for `duration` seconds (-duration is already an existing argument which defaults to 30), checking again later with the daemon for a difficulty drop.

- `-address=<Address>` command-line argument to `bitcoin-miner`. This required RPC support on the server-side.  Basically it does what you would expect -- can mine to any address rather than requiring an address in the daemon wallet.

- RPC support to `getminingcandidate` to support specifying an address. This is an optional argument and required a little bit of parameter passing into `unlimited.cpp` and `mining.cpp` `mkblocktemplate` (see PR).

- Minor performance/efficiency:    Added a mechanism for all threads to realize when prevBlockHash has changed.  When 1 thread finds a block or detects a block change, all threads now know about it. This fixes the situation whereby some threads continued mining on top of old/stale blocks. With this fix, the threads share a global variable (atomically updated) which captures the prev block "cheap hash" as well as the nBits of the latest block template.  If any threads detect this variable has been updated,  they abort working on their current block template as soon as they can and ask the RPC server for a new block right away to not waste time mining on top of a block that is not at chainTip.

- Small performance nit in `bitcoin-miner.cpp` 's `ScanHash` function to not waste CPU cycles needlessly returning from the function every 4096 tries (only to immediately and unconditionally re-enter the function and resume from where it left off).

#### Questions I had and other notes

I had to modify `mining.cpp::mkblocktemplate` a little bit to support:

1. Accepting a `coinbaseScript` argument (if not supplied, does the old codepath of generating wallet address).
2. As before, it must cache its arguments to decide if it should return the same old template or make a new one..  as such, I added some logic to the big `if` statement in `mkblocktemplate` which decides if it should invalidate the cached block template or return a new one -- I hope I got it right. Please look at that section and let me know...
3. This PR actually fixes a weird bug in that `getminingcandidate` would always fail if compiled without wallet support (because it always unconditionally tried to generate to a wallet address, and it would fail if no wallet support) -- so now at least there is a codepath for running `getminingcandidate` without wallet support.

---

#### ~~Note: if you accept this PR, please squash it as it's a noisy one with many small commits.~~

(EDIT: I already squashed this into 1 commit -- can be squash-merged or merge-merged either work now).

If you decide you accept this feature, please let me know if anything needs to be modified here. Thanks!

cculianu commented on src/bitcoin-miner.cpp:
```diff
@@ -76,10 +76,6 @@ bool static ScanHash(const CBlockHeader *pblock, uint32_t &nNonce, uint256 *phas
             return true;
-
-        // If nothing found after trying for a while, return -1
-        if ((nNonce & 0xfff) == 0)
```
Note to anyone reading this: This check was not needed.  `ScanHash` would return here only to be immediately called again unconditionally and continue from where it last left off.  So that's why it was removed.

ptschip commented:
A couple of pointer, not really related to this PR but just in general when submitting PR's.

1) Your PR's are not kicking off the automated travis tests. This is probably because Travis has flagged your account (basically they recently flagged everbody).  You'll need to contact travis support to get your account enabled.

2) Once you get your travis runs re-enabled, be careful not to push every new commit you make up to the PR. This causes a new travis run to kick off which consumes a lot of resources. We already have issues with slow travis runs so it's best to keep them to a minimum by making whatever set of changes you want to make locally and then push all the commit's at one time. It doesn't always work out so perfectly but just be aware that every time you push to your PR your using up travis resources.

And thanks for your submissions, always great to see someone getting involved!

cculianu commented:
Ha, pretty funny with Travis flagging everyone.

1. I will go ahead and contact Travis and have them unflag me. I guess people have been abusing their good graces -- I may have abused Travis in the past (back when I was contributing to Bitcoin ABC we used to push commits willy nilly to kick off test builds a bit excessively; which would explain why I am flagged).  I'll contact them and get my account unblocked.

2. Sorry! I'm a bit commit-happy at times.  I should have actually waited a day or two to submit this PR when it was "ready".  I've been running the miner for 2 days now and observing how it behaves and I keep getting "new ideas".  It feels exciting to put a PR in and then update it -- but if that interferes with Travis and/or your reputation with them (once I get unblocked)  --- I'll just impose a 2-day waiting period and hopefully push PR's as 1 or 2 logical commits.

Thanks a lot for your guidance and taking the time to explain this to me.  I'll go ahead and contact Travis now.  👍

sickpig commented on src/miner.h:
```diff
@@ -151,7 +151,10 @@ int64_t UpdateTime(CBlockHeader *pblock, const Consensus::Params &consensusParam
+UniValue mkblocktemplate(const UniValue &params,
+    int64_t coinbaseSize = -1,
+    CBlock *pblockOut = nullptr,
+    boost::shared_ptr<CReserveScript> coinbaseScript = boost::shared_ptr<CReserveScript>());
```
why not using `std::shared_ptr`? (also on other parts of the PR)

cculianu replied:
Yeah -- I'm not clear on if I made a good decision here.  In fact reading the code again it's hard to defend this decision.

I think I saw other parts of the code do this -- namely the wallet code expected a `boost::shared_ptr<CReserveScript>` and I wanted to be "in line" with that -- but as I read the code it's hard to justify it. In fact, it looks not-ideal, actually. :)

A simple reference to a `CReserveScript` or even a `CScript` would suffice, really.  It's not like that data structure is expensive to copy or this will be called thousands of times per second (and if it were it's arguable the `prevector`-based `CReserveScript` or `CScript` directly is faster/less heap intense).

So.. actually just passing a reference to the `CReserveScript` should be fine... and caching it in the `static` works.

Should I go ahead and do that?

--

Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).

cculianu replied:
Oh -- I know why I did it. In `mining.cpp generateBlocks` ... it actually also takes a `boost::shared_ptr<CReserveScript>` and this code is similar to that and I wanted to respect whatever style/preferences were already in place... but I can change it. Let me know what you recommend!

And.. If you don't mind -- and if you know -- why does that take a `shared_ptr` there?

sickpig replied:
> Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).

dunno, maybe @gandrewstone 	could chime in here. 

My question about using `shared_ptr` was more related to the fact that we are trying to move away from boost (even thou slowly) and I didn't realize you chose `boost::shared_ptr` because `generateBlocks` use this type for one of its param

cculianu replied:
Oh .. I see.. the wallet expects a pointer because it may return a polymorphic type (virtual methods to `CReserveScript`) -- which makes me wonder -- does this codepath need to call `KeepKey()` after generating a block with the wallet address in coinbase? (Note this PR did not introduce this concern -- it would have been there before this PR anyway).  I'm actually not clear on what `KeepKey` does in the wallet .. I can read the `wallet.cpp` code I guess. :)

cculianu replied:
> > Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).
> 
> dunno, maybe @gandrewstone could chime in here.
> 
> My question about using `shared_ptr` was more related to the fact that we are trying to move away from boost (even thou slowly) and I didn't realize you chose `boost::shared_ptr` because `generateBlocks` use this type for one of its param

I agree that using the `std::` variety (which is basically boost anyway) is the way to go now that C++14, C++17, etc are here. I guess using the `boost` parts not in `std` would be ideal..

Yeah I would like to know what to do here -- I really just wanted to "behave" like the other code path -- note that privately a `boost::shared_ptr` would still be needed here to grab a public key from the wallet in the case where caller didn't pass one down....

I'll do whatever you guys say.  Any guidance is appreciated here. I'm new to the codebase (well, it's an old familiar friend as it resembles what I worked on in the past a great deal, with some minor and major variation.. :) ).

gandrewstone replied:
Please use references when the function arg's life is short (as short as the function).  The reference will be protected from deletion by the caller's context.

Use the c++14 std:: variety whenever possible, and you can convert some boost:: equivalents opportunistically.  However, if enough conversions pile up to obscure the point of the PR, they should be isolated in their own PR.

gandrewstone commented:
I like the direction you are going here!  But I need to focus on #1857 first and I'm not sure how long that will take before looking at this in detail.  No more than a few days...

cculianu commented:
Deadlock detector -- ooh. Sounds great.  Thanks for the status update and looking forward to any and all insights for this PR, whenever you can.  :)

dagurval commented on src/rpc/mining.cpp:
```diff
@@ -678,25 +682,50 @@ UniValue mkblocktemplate(const UniValue &params, int64_t coinbaseSize, CBlock *p
+        (consensusParams.fPowAllowMinDifficultyBlocks && std::abs(GetTime() - nStart) > 30) || // 3 above
+        // 4 above
+        (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLast && std::abs(GetTime() - nStart) > 5) ||
+        prevCoinbaseSize != coinbaseSize || !!prevCoinbaseScript != !!coinbaseScript || // 5 above
```
nit: `bool(stuff)` is more readable than `!!stuff`

cculianu replied:
Ok, will change.

dagurval commented on src/util.cpp:
```diff
@@ -568,6 +568,13 @@ int64_t GetArg(const std::string &strArg, int64_t nDefault)
+double GetDoubleArg(const std::string &strArg, double dDefault)
+{
+    if (mapArgs.count(strArg))
+        return atof(mapArgs[strArg].c_str()); // returns 0.0 on conversion failure
```
Value on conversion failure is consistent with `int64_t GetArg`. Hiding a conversion error and just returning some value is a bad convention though, but out of scope for this PR I think.

cculianu replied:
I agree -- the preferred pattern I usually like is passing an optional `bool * = nullptr` arg and if that's present, writing the success/error result of the conversion there. If you like I can add a PR after this one that does that for all the `GetArg` functions in this codebase.. it would be easy/fun to do so... :)

dagurval replied:
That's a good proposal and I would definitely support it (I don't have merge privileges though, so can't promise anything).

gandrewstone replied:
I'll merge that PR!

gandrewstone replied:
... or throw an exception? ...your choice

dagurval commented on src/unlimited.cpp:
```diff
@@ -1759,18 +1759,52 @@ UniValue getminingcandidate(const UniValue &params, bool fHelp)
+        catch (...)
+        {
+            // parse failure on coinbaseSize, must be address
+            destStr = params[0].get_str();
```
This is not how optional positional parameters work 🤔

cculianu replied:
Indeed. Note that I modified the PR now to accept:

`getminincandidate` # no coinbaseSize specified, use default address (daemon wallet)
`getminingcandidate 1024` # specify custom coinbaseSize
`getminingcandidate 1024 bchtest:bla...` # use custom coinbaseSize, custom address
`getminingcandidate null bchtest:bla...` # use **default** coinbaseSize, custom address
^ This is new

I hope this is more in line with how things are expected to work. Note the only change here to the RPC API would be before `null` was rejected as a coinbaseSize (it had no meaning) -- and now it means "use default from daemon tweaks".

dagurval reviewed (commented):
I disagree with allowing address to be *either* first OR seconds argument to `getminingcandidate` (ping @gandrewstone). The rest looks good!

cculianu commented:
> I disagree with allowing address to be _either_ first OR seconds argument to `getminingcandidate` (ping @gandrewstone). The rest looks good!

Fair enough. I was "being clever" here but this is different from the way the rest  of the bitcoind RPC works (and perhaps unexpected and potentially annoying to other devs -- because of that). You are right in that generally optional positional arguments always retain their position index.  So for instance here it should be:

No args:

`[]`

One optional arg set to 1024:

`[1024]`

Both args set:

`[ 1024, "bchtest:qqsdas.." ]`

Just setting the second arg (first one should be 0 which should be interpreted as unset):

`[ 0, "bchtest:qqwesads..." ]`

So I can modify this code to do the above -- would that be acceptable?

sickpig committed:
Fix bash syntax error in .travis.yml

dagurval reviewed (approved):
LGTM. I also like the new optional argument approach to `getminingcandidate`.

cculianu committed:
CPU Miner Enchancements: add -address=, -maxdifficulty= and other tweaks

I have been mining testnet lately (hashrate seems to be mostly CPU now
anyway on testnet, with the occasional ASIC miner showing up to rain
blocks for 12 hours before leaving abruptly).

As such, I have been using the `bitcoin-miner` program quite a bit and I
wanted to add a few features to it to make it more useful for everybody.

This commit adds the following:

- `-maxdifficulty=<float>` command-line argument to `bitcoin-miner`. With
this option, the miner doesn't bother mining if difficulty is too high.
It's pointless to mine with a CPU when difficulty is too high (extra wear
& tear on the hardware).  Whenever difficulty is > `maxdifficulty` it just
sleeps for `duration` seconds (-duration is already an existing argument
which defaults to 30), checking again later with the daemon for a
difficulty drop.

- `-address=<Address>` command-line argument to `bitcoin-miner`. This
required RPC support on the server-side.  Basically it does what you would
expect -- can mine to any address rather than requiring an address in the
daemon wallet.

- RPC support to `getminingcandidate` to support specifying an address.
This is an optional argument and required a little bit of parameter passing
into `unlimited.cpp` and `mining.cpp` `mkblocktemplate` (see PR).

- Minor performance/efficiency:    Added a mechanism for all threads to
realize when prevBlockHash has changed.  When 1 thread finds a block or
detects a block change, all threads now know about it. This fixes the
situation whereby some threads continued mining on top of old/stale blocks.
With this fix, the threads share a global variable (atomically updated)
which captures the prev block "cheap hash" as well as the nBits of the
latest block template.  If any threads detect this variable has been
updated,  they abort working on their current block template as soon as
they can and ask the RPC server for a new block right away to not waste
time mining on top of a block that is not at chainTip.

- Small performance nit in `bitcoin-miner.cpp` 's `ScanHash` function to
not waste CPU cycles needlessly returning from the function every 4096
tries (only to immediately and unconditionally re-enter the function and
resume from where it left off).

cculianu commented:
ARGH! I force-pushed a squashed commit (1 commit) of all this stuff and github hates that.

sickpig commented:
Travis failure is unrelated to this PR. 

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/583586666

@cculianu if you rebase on top of #1910 the error will go away.

cculianu commented:
@sickpig Ok, I rebased on top of that commit 3f65600e04c367365455fe6ce531541f65b8d5a3. It started the build again.. hopefully will work.

gandrewstone commented:
Thanks for this work!

cculianu commented:
My pleasure!

Oh hey I was going to clean this up as per your recommendations... regarding references.  I am just on mobile now..


I’ll just spam another PR for that later if you like.. :)
